Add ScheduleAt ability.

Bluebugs opened the pull request:
This should address #19

Commit:
Add ScheduleAt ability.

Bluebugs commented:
This should addres #19.

Bluebugs commented:
I think this PR should help @jtdaad schedule update during office off hours.

andydotxyz commented on updater.go:
```diff
@@ -29,9 +29,25 @@ type Config struct {
+
+const (
+	None    Repeating = iota
+	Minutly           // Will schedule in the next minute and repeat it every minute after
```
Can we remove this? No application should try and update every minute!

andydotxyz commented on cmd/example/main.go:
```diff
@@ -17,8 +17,15 @@ func main() {
+		Source: httpSource,
+		Schedule: selfupdate.Schedule{
+			FetchOnStart: true,
+			Interval:     time.Minute * time.Duration(60),
```
I don't know if this is a good example - as the "ScheduleAt" will happen in addition to every hour after start...

andydotxyz commented on updater.go:
```diff
@@ -121,14 +137,37 @@ func Manage(conf *Config) (*Updater, error) {
+					now := time.Now()
+					var next time.Time
+					switch updater.conf.Schedule.At.Repeating {
+					case Minutly:
```
This really should be factored out into another method that determines what the "next time" for a schedule.
That can also then be tested - this is some pretty clever date manipulation that has no tests.

andydotxyz commented on updater.go:
```diff
@@ -121,14 +137,37 @@ func Manage(conf *Config) (*Updater, error) {
+						next = time.Date(now.Year(), now.Month()+1, updater.conf.Schedule.At.Day(), updater.conf.Schedule.At.Hour(), updater.conf.Schedule.At.Minute(), updater.conf.Schedule.At.Second(), updater.conf.Schedule.At.Nanosecond(), time.Local)
+					}
+					time.Sleep(next.Sub(now))
+					updater.CheckNow()
```
Ignored error

andydotxyz commented on updater.go:
```diff
@@ -121,14 +137,37 @@ func Manage(conf *Config) (*Updater, error) {
-				err := updater.CheckNow()
-				if err != nil {
-					logError("Upgrade error: %v\n", err)
+			go func() {
```
This is not too cool, we now have 2 goroutines spawning from within a goroutine none of which has date tests.
This would be so much cleaner if this single goroutine was able to get the next time to trigger from a helper (tested) function, then we don't need 3 goroutines and 2 separate inner for loops.

andydotxyz reviewed (changes requested):
Complexity is drastically increased here - see comments inline for how it could be improved.
If we think test first it will be clear that this code needs to be tidied.

Commit:
Reduce complexity and only use one nested go func.

Commit:
We should honor the offset timezone.

This should allow flexibility to easily synchronize with remote server
for example.

Commit:
Try to make example less confusing, but still cover both scenario.

andydotxyz commented on cmd/example/main.go:
```diff
@@ -20,11 +20,12 @@ func main() {
 		Schedule: selfupdate.Schedule{
 			FetchOnStart: true,
-			Interval:     time.Minute * time.Duration(60),
+			// If you want to check update on a regular interval uncomment the follwoing
```
"following"

andydotxyz commented on updater.go:
```diff
@@ -203,3 +174,41 @@ func applyUpdate(r io.Reader, publicKey ed25519.PublicKey, signature [64]byte) (
+
+func triggerSchedule(updater *Updater) {
+	for {
+		var next time.Duration
```
Sorry to be picky, but `next` is a time not a duration, perhaps "offset" or "delta" or "delay" etc might work better?

andydotxyz commented on updater.go:
```diff
@@ -203,3 +174,41 @@ func applyUpdate(r io.Reader, publicKey ed25519.PublicKey, signature [64]byte) (
+	}
+}
+
+func nextScheduleAt(repeating Repeating, offset time.Time) time.Duration {
```
Much like the comment above this name implies returning a time but it returns a duration.
Perhaps "delayUntilNextTrigger" or "offsetToSchedule" etc?

andydotxyz reviewed (changes requested):
Much cleaner thanks, just a little misleading naming to resolve.

Commit:
Correct typo and naming.

andydotxyz reviewed (commented):
Looking good, but some missing docs.
I have turned lint on in main now, so it should highlight them

andydotxyz reviewed (approved):
Cedric promised the docs in next PR :)
